db/search/search.go

```diff
@@ -0,0 +1,36 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package search
+
+import (
+	"strings"
+)
+
+const (
+	LIKE = "like"
+)
+
+type Search struct {
+	Value string
+}
+
+func NewSearchField(value string) Search {
+	searchValue := strings.Builder{}
```

Comment:
this comment https://github.com/cloudwan/gohan/pull/731#discussion_r263075188 is still valid and needs to be resolved

Reply:
In case we are escaping the characters '%','_','/' with slash or something, we need to make sure that when the query is executed the escape character is handled in the queries 
i.e for instance with maria DB
https://mariadb.com/kb/en/library/like/ you specify the escape character while executing like queries

otherwise `%`+'/'+`%` can be executed as `%`+'//'+`%` and might result in incorrect results if not handled.

Since we are using Masterminds/squirrel package i don't see a place where escape characters are handled while framing the query strings

https://github.com/Masterminds/squirrel/blob/55303df43ec0bf2b0ce3f9c3d8c3ce07022ef76b/expr.go#L169

Reply:
By default, escape character is set to backslash. 'squirrel' doesn't have to support changing it, we can go with backslash.
https://mariadb.com/kb/en/library/like/
> If you need to match the characters _ or %, you must escape them. By default, you can prefix the wildcard characters the backslash character \ to escape them. The backslash is used both to encode special characters like newlines when a string is parsed as well as to escape wildcards in a pattern after parsing. Thus, to match an actual backslash, you sometimes need to double-escape it as "\\\\".

Reply:
yes...mariadb supports backslash...gohan is DB agnostic right..isn't that the reason why squirrel is used ....don't we need to handle for other databases as well.. I tested with sqlite and it yielded wrong results, when escaped with backlash..

Reply:
Well, in this case you can always implement you own Where matcher similar to existing sq.Like, which supports ESCAPE clause. All you have to do is to implement sq.Sqlizer interface on your custom struct.

Reply:
implementing sq.Sqlizer sound a bit like an overkill for me. maybe I'm missing something, but can we implement [driver.Valuer](https://github.com/Masterminds/squirrel/blob/v1/expr.go#L183) instead and stay with regular sq.Like? eg.

```Go
func (s *Search) Value() (driver.Value, error) {
	buf := strings.Builder{}
	buf.WriteString("%")
	buf.WriteString(escape(s.value))
	buf.WriteString("%")
	return buf.String()
}
```

Reply:
apart from this suggested change..i escaped characters and ran test with the server_test.go..still it fails in producing the correct results as i believe the test configuration is not mysql..so tests for the search feature fails with escaped chars..which brings to ..why aren't we testing the api's on a set of databases..rather than just sqlite3

Reply:
@p-kozlowski As far as I understand, sqlite doesn't support backslash escape character by default. To enable it, query must contain optional ESCAPE statement followed by escape character. In MariaDB, backslash is enabled by default. squirrel doesn't provide such query modifier, but exposes interace which could be implemented to add such field. 

@giyyanan i see that characters are escaped, and you wrote that some tests are failing. Those on CircleCI are passing. Does CI skips any tests?

Reply:
I haven't checked in the code for unit test cases with special characters..